File: conformance/bundles/16-stimulus-state-replacement/ReplaceSteps.java
package dev.varar.conformance.bundle16;

import dev.varar.State;
import dev.varar.StepDefinitions;
import dev.varar.Steps;

/**
 * Java sibling of {@code replace.steps.ts} (bundle {@code 16-stimulus-state-replacement}).
 *
 * <p>The second stimulus builds a {@code Ctx} carrying only {@code b}; {@code a} falls back to 0.
 * The record shape makes full replacement the only expressible contract here — which is precisely
 * why this bundle's golden pins the dynamic ports (TypeScript, Python, Ruby) to the same answer.
 */
public final class ReplaceSteps implements StepDefinitions<ReplaceSteps.Ctx> {

    record Ctx(int a, int b) implements State {}

    @Override
    public void register(Steps<Ctx> s) {
        s.defineState(() -> new Ctx(0, 0));

        s.stimulus("I set a to 1 and b to 2", (Ctx ctx) -> new Ctx(1, 2));

        s.stimulus("I set only b to 3", (Ctx ctx) -> new Ctx(0, 3));

        s.sensor("Then a is {int} and b is {int}", (Ctx ctx, Integer a, Integer b) ->
                java.util.List.of(ctx.a(), ctx.b()));
    }
}

File: conformance/bundles/16-stimulus-state-replacement/replace.steps.rs
//! Rust sibling of `replace.steps.ts` (bundle `16-stimulus-state-replacement`).
//!
//! The second stimulus builds a `Ctx` carrying only `b`; `a` falls back to 0.
//! The struct shape makes full replacement the only expressible contract here —
//! which is why this bundle's golden pins the dynamic ports to the same answer.

use varar::Steps;

#[derive(Clone, Default)]
pub struct Ctx {
    pub a: i64,
    pub b: i64,
}

pub fn register(s: &mut Steps<Ctx>) {
    s.stimulus("I set a to 1 and b to 2", |_ctx: Ctx| Ok(Ctx { a: 1, b: 2 }));
    s.stimulus("I set only b to 3", |_ctx: Ctx| {
        Ok(Ctx {
            b: 3,
            ..Default::default()
        })
    });
    s.sensor("Then a is {int} and b is {int}", |ctx: Ctx, _a: i64, _b: i64| Ok((ctx.a, ctx.b)));
}

pub fn state() -> Ctx {
    Ctx::default()
}
